# 6.4 Cycle Data Munging

***Declaration ***: The coding  is used was abstract  from Kevin mark ham youtube video seriese, Introduction to machine learning with scikit-learn video series. You can find link under resources section.  

# 6.4.1 Data Munging steps
1.	Handling missing Data
2.	Encoding, Decoding, and recoding of data
3.	Handling Anomalous values
4.	Transforming Data
5.	Merging Dataset



# Data Cleaning

In [2]:
# load libraries and set styles, options
import os,csv
import numpy as np
import pandas as pd
import seaborn as sns
from IPython.display import HTML
import warnings; warnings.simplefilter('ignore')

In [3]:
%matplotlib inline

In [4]:
# read in a CSV
df = pd.read_csv('C:/Users/mrferozi/Desktop/GitHub/Bike/cycle-share-dataset-orignal/trip.csv')

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 286857 entries, 0 to 286856
Data columns (total 12 columns):
trip_id              286857 non-null int64
starttime            286857 non-null object
stoptime             286857 non-null object
bikeid               286857 non-null object
tripduration         286857 non-null float64
from_station_name    286857 non-null object
to_station_name      286857 non-null object
from_station_id      286857 non-null object
to_station_id        286857 non-null object
usertype             286857 non-null object
gender               181557 non-null object
birthyear            181553 non-null float64
dtypes: float64(2), int64(1), object(9)
memory usage: 26.3+ MB


The data set contains trip duration for less than three minutes, which could be possible because it may happen that the user took the bike from cycle dock and decide not to proceed with the journey and the user return bike to the dock. Another, thing which we are ignoring for the moment is the journey recorded for the short pass holder which does not kick in any meaning in our current analysis, Note, a separate analysis has performed for short pass holders but  at this point this study decide to get rid of undesirable data and more focus on the data which contributing significance in our analysis.

In [5]:
# Filter data and only consider trip duration more than 3 minute and exclude Short-Term Pass Holder from our data set
bikes = df[(df.tripduration >=300) & (df.usertype != 'Short-Term Pass Holder')]

In [6]:
bikes.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 143543 entries, 0 to 286847
Data columns (total 12 columns):
trip_id              143543 non-null int64
starttime            143543 non-null object
stoptime             143543 non-null object
bikeid               143543 non-null object
tripduration         143543 non-null float64
from_station_name    143543 non-null object
to_station_name      143543 non-null object
from_station_id      143543 non-null object
to_station_id        143543 non-null object
usertype             143543 non-null object
gender               143543 non-null object
birthyear            143541 non-null float64
dtypes: float64(2), int64(1), object(9)
memory usage: 14.2+ MB


- After applying fillter our data set shrink from 286857 to 143541 tuples

#  Imputation (Handle missing values)

**What does "NaN" mean?**

- "NaN" is not a string, rather it's a special value: **`numpy.nan`**.
- It stands for "Not a Number" and indicates a **missing value**.
- **`read_csv`** detects missing values (by default) when reading the file, and replaces them with this special value.

In [7]:
# 'isnull' returns a DataFrame of booleans (True if missing, False if not missing)
bikes.isnull().tail()

trip_id starttime stoptime bikeid tripduration from_station_name  \
286842   False     False    False  False        False             False   
286843   False     False    False  False        False             False   
286844   False     False    False  False        False             False   
286845   False     False    False  False        False             False   
286847   False     False    False  False        False             False   

       to_station_name from_station_id to_station_id usertype gender birthyear  
286842           False           False         False    False  False     False  
286843           False           False         False    False  False     False  
286844           False           False         False    False  False     False  
286845           False           False         False    False  False     False  
286847           False           False         False    False  False     False

In [8]:
# 'nonnull' returns the opposite of 'isnull' (True if not missing, False if missing)
bikes.notnull().tail()

trip_id starttime stoptime bikeid tripduration from_station_name  \
286842    True      True     True   True         True              True   
286843    True      True     True   True         True              True   
286844    True      True     True   True         True              True   
286845    True      True     True   True         True              True   
286847    True      True     True   True         True              True   

       to_station_name from_station_id to_station_id usertype gender birthyear  
286842            True            True          True     True   True      True  
286843            True            True          True     True   True      True  
286844            True            True          True     True   True      True  
286845            True            True          True     True   True      True  
286847            True            True          True     True   True      True

In [9]:
# count the number of missing values in each Series
bikes.isnull().sum()

trip_id              0
starttime            0
stoptime             0
bikeid               0
tripduration         0
from_station_name    0
to_station_name      0
from_station_id      0
to_station_id        0
usertype             0
gender               0
birthyear            2
dtype: int64

**handle missing values** depends on the dataset as well as the nature of analysis

In [10]:
bikes.shape

(143543, 12)

In [11]:
# if 'any' values are missing in a row, then drop that row
bikes.dropna(how='any').shape

(143541, 12)

In [12]:
bikes.birthyear.mean()

1979.3765474672741

In [13]:
# fill in missing values with a specified value
bikes['birthyear'].fillna(value='1979', inplace=True)

In [14]:
# count the number of missing values in each Series
bikes.isnull().sum()

trip_id              0
starttime            0
stoptime             0
bikeid               0
tripduration         0
from_station_name    0
to_station_name      0
from_station_id      0
to_station_id        0
usertype             0
gender               0
birthyear            0
dtype: int64

# Data Encoding

converting data Categorical to Numaric for further use

- Converting Gender Column 

In [15]:
# create the 'Sex_num' dummy variable using the 'map' method
bikes['Sex_num'] = bikes.gender.map({'Female':2, 'Male':1})

- Converting category From Station ID and To Station ID to nominal

In [16]:
bikes["from_station_id_cat"] = bikes["from_station_id"].astype('category')
bikes.dtypes 

trip_id                   int64
starttime                object
stoptime                 object
bikeid                   object
tripduration            float64
from_station_name        object
to_station_name          object
from_station_id          object
to_station_id            object
usertype                 object
gender                   object
birthyear                object
Sex_num                 float64
from_station_id_cat    category
dtype: object

In [17]:
bikes["from_station_id_num"] = bikes["from_station_id_cat"].cat.codes

In [18]:
bikes["to_station_id_cat"] = bikes["to_station_id"].astype('category')
bikes.dtypes 

trip_id                   int64
starttime                object
stoptime                 object
bikeid                   object
tripduration            float64
from_station_name        object
to_station_name          object
from_station_id          object
to_station_id            object
usertype                 object
gender                   object
birthyear                object
Sex_num                 float64
from_station_id_cat    category
from_station_id_num        int8
to_station_id_cat      category
dtype: object

In [19]:
bikes["to_station_id_num"] = bikes["to_station_id_cat"].cat.codes

# Data Encoding

For further, for our analysis we are dividing date columns in small portion. Which, could help us to understand dataset more clearly. 

In [20]:
# convert 'Time' to datetime format
bikes['starttime'] = pd.to_datetime(bikes.starttime)

In [21]:
bikes.dtypes

trip_id                         int64
starttime              datetime64[ns]
stoptime                       object
bikeid                         object
tripduration                  float64
from_station_name              object
to_station_name                object
from_station_id                object
to_station_id                  object
usertype                       object
gender                         object
birthyear                      object
Sex_num                       float64
from_station_id_cat          category
from_station_id_num              int8
to_station_id_cat            category
to_station_id_num                int8
dtype: object

In [22]:
bikes['Day'] = bikes.starttime.dt.weekday_name

In [23]:
# convert 'Time' to datetime format
bikes['stoptime'] = pd.to_datetime(bikes.stoptime)

In [24]:
bikes["Day_cat"] = bikes["Day"].astype('category')
bikes.dtypes 

trip_id                         int64
starttime              datetime64[ns]
stoptime               datetime64[ns]
bikeid                         object
tripduration                  float64
from_station_name              object
to_station_name                object
from_station_id                object
to_station_id                  object
usertype                       object
gender                         object
birthyear                      object
Sex_num                       float64
from_station_id_cat          category
from_station_id_num              int8
to_station_id_cat            category
to_station_id_num                int8
Day                            object
Day_cat                      category
dtype: object

In [25]:
bikes["Day_num"] = bikes["Day_cat"].cat.codes

In [26]:
# convenient Series attributes are now available
bikes["sthours"] = bikes.starttime.dt.hour

In [27]:
# convenient Series attributes are now available
bikes["stphours"] = bikes.stoptime.dt.hour

In [28]:
bikes['tripduration_minutes']=bikes['tripduration']/60
desred_decimals = 2    
bikes['tripduration_minutes'] = bikes['tripduration_minutes'].apply(lambda x: round(x,desred_decimals))

In [29]:
 bikes['birthyear'] = bikes['birthyear'].astype(int)

In [30]:
bikes['age'] = 2018 - bikes['birthyear']
desred_decimals = 2    
bikes['birthyear'] = bikes['birthyear'].apply(lambda x: round(x,desred_decimals))
bikes.age.head()

0    58
1    48
2    30
3    41
4    47
Name: age, dtype: int32

Some statistics about age Variable 

In [31]:
bikes.age.describe()

count    143543.000000
mean         38.623458
std          10.251359
min          19.000000
25%          31.000000
50%          36.000000
75%          44.000000
max          87.000000
Name: age, dtype: float64

In [32]:
bikes.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 143543 entries, 0 to 286847
Data columns (total 24 columns):
trip_id                 143543 non-null int64
starttime               143543 non-null datetime64[ns]
stoptime                143543 non-null datetime64[ns]
bikeid                  143543 non-null object
tripduration            143543 non-null float64
from_station_name       143543 non-null object
to_station_name         143543 non-null object
from_station_id         143543 non-null object
to_station_id           143543 non-null object
usertype                143543 non-null object
gender                  143543 non-null object
birthyear               143543 non-null int64
Sex_num                 140905 non-null float64
from_station_id_cat     143543 non-null category
from_station_id_num     143543 non-null int8
to_station_id_cat       143543 non-null category
to_station_id_num       143543 non-null int8
Day                     143543 non-null object
Day_cat                 143

# Removing Duplicate values

In [33]:
# detect duplicate trip_id codes: True if an item is identical to a previous item
bikes.trip_id.duplicated().tail()

286842    False
286843    False
286844    False
286845    False
286847    False
Name: trip_id, dtype: bool

The above codes shows that we do not have any duplicated rows in our dataset

In [34]:
# Breaking date column for further Analysis
bikes['bmonth'] = bikes['starttime'].dt.month

In [35]:
# Breaking date column for further Analysis
bikes['Date'] = bikes['starttime'].dt.date

In [36]:
bikes.dtypes

trip_id                          int64
starttime               datetime64[ns]
stoptime                datetime64[ns]
bikeid                          object
tripduration                   float64
from_station_name               object
to_station_name                 object
from_station_id                 object
to_station_id                   object
usertype                        object
gender                          object
birthyear                        int64
Sex_num                        float64
from_station_id_cat           category
from_station_id_num               int8
to_station_id_cat             category
to_station_id_num                 int8
Day                             object
Day_cat                       category
Day_num                           int8
sthours                          int64
stphours                         int64
tripduration_minutes           float64
age                              int32
bmonth                           int64
Date                     

In [37]:
bikes.Day_num.tail()

286842    6
286843    6
286844    6
286845    6
286847    6
Name: Day_num, dtype: int8

# Filtering Data for year 2015 and 2016

In [38]:
# convert 'Time' to datetime format
bikes['starttime'] = pd.to_datetime(bikes.starttime)

In [39]:
# Breaking date column for further Analysis
bikes['year'] = bikes['starttime'].dt.year

In [40]:
# convert 'Time' to datetime format
bikes['Date'] = pd.to_datetime(bikes.Date)
bikes.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 143543 entries, 0 to 286847
Data columns (total 27 columns):
trip_id                 143543 non-null int64
starttime               143543 non-null datetime64[ns]
stoptime                143543 non-null datetime64[ns]
bikeid                  143543 non-null object
tripduration            143543 non-null float64
from_station_name       143543 non-null object
to_station_name         143543 non-null object
from_station_id         143543 non-null object
to_station_id           143543 non-null object
usertype                143543 non-null object
gender                  143543 non-null object
birthyear               143543 non-null int64
Sex_num                 140905 non-null float64
from_station_id_cat     143543 non-null category
from_station_id_num     143543 non-null int8
to_station_id_cat       143543 non-null category
to_station_id_num       143543 non-null int8
Day                     143543 non-null object
Day_cat                 143

In [41]:
# df2 belong to bikes dataset
df2 = bikes[(bikes.year != 2014) ]
df2.head(100)

trip_id           starttime            stoptime    bikeid  \
20240    25092 2015-01-01 00:37:00 2015-01-01 00:44:00  SEA00267   
20252    25114 2015-01-01 03:07:00 2015-01-01 03:18:00  SEA00472   
20255    25127 2015-01-01 08:12:00 2015-01-01 08:20:00  SEA00245   
20260    25132 2015-01-01 10:55:00 2015-01-01 11:03:00  SEA00391   
20263    25135 2015-01-01 11:31:00 2015-01-01 11:55:00  SEA00079   
20265    25137 2015-01-01 11:45:00 2015-01-01 11:54:00  SEA00401   
20270    25143 2015-01-01 12:13:00 2015-01-01 12:22:00  SEA00347   
20271    25144 2015-01-01 12:14:00 2015-01-01 12:22:00  SEA00326   
20274    25147 2015-01-01 12:16:00 2015-01-01 12:27:00  SEA00140   
20276    25150 2015-01-01 12:52:00 2015-01-01 13:16:00  SEA00212   
20277    25151 2015-01-01 12:52:00 2015-01-01 12:59:00  SEA00390   
20281    25155 2015-01-01 12:58:00 2015-01-01 13:16:00  SEA00210   
20282    25156 2015-01-01 12:59:00 2015-01-01 13:16:00  SEA00141   
20292    25168 2015-01-01 13:17:00 2015-01-01 13:32:00  SEA00255   
20325    25203 2015-01-01 14:22:00 2015-01-01 14:31:00  SEA00390   
20327    25205 2015-01-01 14:27:00 2015-01-01 14:46:00  SEA00212   
20330    25210 2015-01-01 14:49:00 2015-01-01 14:57:00  SEA00319   
20331    25211 2015-01-01 15:00:00 2015-01-01 15:15:00  SEA00264   
20332    25212 2015-01-01 15:02:00 2015-01-01 15:12:00  SEA00206   
20333    25213 2015-01-01 15:15:00 2015-01-01 15:25:00  SEA00418   
20337    25217 2015-01-01 15:19:00 2015-01-01 15:27:00  SEA00056   
20342    25222 2015-01-01 15:27:00 2015-01-01 15:36:00  SEA00312   
20346    25226 2015-01-01 15:40:00 2015-01-01 15:57:00  SEA00215   
20347    25228 2015-01-01 15:46:00 2015-01-01 15:54:00  SEA00056   
20348    25229 2015-01-01 15:45:00 2015-01-01 15:58:00  SEA00382   
20350    25231 2015-01-01 16:04:00 2015-01-01 16:12:00  SEA00212   
20351    25232 2015-01-01 16:04:00 2015-01-01 16:13:00  SEA00135   
20356    25237 2015-01-01 16:17:00 2015-01-01 16:33:00  SEA00421   
20357    25238 2015-01-01 16:19:00 2015-01-01 16:27:00  SEA00278   
20358    25239 2015-01-01 16:23:00 2015-01-01 16:28:00  SEA00116   
...        ...                 ...                 ...       ...   
20437    25324 2015-01-02 10:42:00 2015-01-02 11:04:00  SEA00299   
20438    25325 2015-01-02 11:21:00 2015-01-02 11:34:00  SEA00100   
20439    25326 2015-01-02 11:25:00 2015-01-02 11:32:00  SEA00309   
20440    25327 2015-01-02 11:38:00 2015-01-02 11:55:00  SEA00168   
20443    25330 2015-01-02 11:52:00 2015-01-02 11:58:00  SEA00208   
20446    25333 2015-01-02 12:13:00 2015-01-02 12:20:00  SEA00405   
20447    25334 2015-01-02 12:21:00 2015-01-02 12:32:00  SEA00277   
20448    25335 2015-01-02 12:26:00 2015-01-02 12:49:00  SEA00153   
20449    25336 2015-01-02 12:26:00 2015-01-02 12:43:00  SEA00496   
20450    25337 2015-01-02 12:39:00 2015-01-02 12:44:00  SEA00340   
20451    25338 2015-01-02 12:43:00 2015-01-02 12:49:00  SEA00198   
20453    25340 2015-01-02 12:50:00 2015-01-02 13:01:00  SEA00196   
20454    25341 2015-01-02 12:52:00 2015-01-02 13:09:00  SEA00496   
20455    25342 2015-01-02 12:52:00 2015-01-02 12:59:00  SEA00056   
20457    25344 2015-01-02 12:56:00 2015-01-02 13:03:00  SEA00309   
20458    25345 2015-01-02 13:02:00 2015-01-02 13:15:00  SEA00100   
20459    25346 2015-01-02 13:08:00 2015-01-02 13:28:00  SEA00447   
20461    25349 2015-01-02 13:16:00 2015-01-02 13:26:00  SEA00135   
20464    25353 2015-01-02 13:26:00 2015-01-02 13:31:00  SEA00483   
20465    25354 2015-01-02 13:38:00 2015-01-02 13:55:00  SEA00063   
20467    25356 2015-01-02 13:56:00 2015-01-02 14:03:00  SEA00232   
20468    25357 2015-01-02 14:03:00 2015-01-02 14:09:00  SEA00356   
20469    25358 2015-01-02 14:10:00 2015-01-02 14:16:00  SEA00278   
20470    25359 2015-01-02 14:12:00 2015-01-02 14:33:00  SEA00477   
20474    25363 2015-01-02 14:32:00 2015-01-02 14:49:00  SEA00168   
20476    25365 2015-01-02 14:48:00 2015-01-02 14:57:00  SEA00235   
20477    25366 2015-01-02 14:53:00 2015-01-02 15:07:00  SEA0047

# Saving Bike data

In [42]:
df2.to_csv('C:\\Users\\mrferozi\\Desktop\\GitHub\\Bike\\dataset\\cycle\\trip_clean.csv')

-----------------------------------------------------

# Cleaning Weather Data

In [5]:
# read in a CSV
df = pd.read_csv('C:/Users/mrferozi/Desktop/GitHub/Bike/cycle-share-dataset-orignal/weather.csv')

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 689 entries, 0 to 688
Data columns (total 21 columns):
Date                          689 non-null object
Max_Temperature_F             689 non-null int64
Mean_Temperature_F            688 non-null float64
Min_TemperatureF              689 non-null int64
Max_Dew_Point_F               689 non-null int64
MeanDew_Point_F               689 non-null int64
Min_Dewpoint_F                689 non-null int64
Max_Humidity                  689 non-null int64
Mean_Humidity                 689 non-null int64
Min_Humidity                  689 non-null int64
Max_Sea_Level_Pressure_In     689 non-null float64
Mean_Sea_Level_Pressure_In    689 non-null float64
Min_Sea_Level_Pressure_In     689 non-null float64
Max_Visibility_Miles          689 non-null int64
Mean_Visibility_Miles         689 non-null int64
Min_Visibility_Miles          689 non-null int64
Max_Wind_Speed_MPH            689 non-null int64
Mean_Wind_Speed_MPH           689 non-null int64
Max_

# Converting temprature from Fahrenheit to Celsius

The definition of precipitation is any form of water - liquid or solid - falling from the sky. It includes rain, sleet, snow, hail and drizzle plus a few less common occurrences such as ice pellets, diamond dust and freezing rain.
https://www.metoffice.gov.uk/learning/learn-about-the-weather/weather-phenomena/what-is-precipitation

- Converting  temprature from Fahrenheit to Celsius

T(°C) = (T(°F) - 32) × 5/9

Example

Convert 68 degrees Fahrenheit to degrees Celsius:


T(°C) = (68°F - 32) × 5/9 = 20 °C
https://www.rapidtables.com/convert/temperature/how-fahrenheit-to-celsius.html

In [7]:
a = df.Mean_Temperature_F-32

In [8]:
df['Mean_Temperature_C'] = a * 5/9
desred_decimals = 2    
df['Mean_Temperature_C'] = df['Mean_Temperature_C'].apply(lambda x: round(x,desred_decimals))

In [9]:
# 'isnull' returns a DataFrame of booleans (True if missing, False if not missing)
df.isnull().tail()

Date Max_Temperature_F Mean_Temperature_F Min_TemperatureF  \
684  False             False              False            False   
685  False             False              False            False   
686  False             False              False            False   
687  False             False              False            False   
688  False             False              False            False   

    Max_Dew_Point_F MeanDew_Point_F Min_Dewpoint_F Max_Humidity Mean_Humidity  \
684           False           False          False        False         False   
685           False           False          False        False         False   
686           False           False          False        False         False   
687           False           False          False        False         False   
688           False           False          False        False         False   

    Min_Humidity        ...         Min_Sea_Level_Pressure_In  \
684        False        ...                             False   
685        False        ...                             False   
686        False        ...                             False   
687        False        ...                             False   
688        False        ...                             False   

    Max_Visibility_Miles Mean_Visibility_Miles Min_Visibility_Miles  \
684                False                 False                False   
685                False                 False                False   
686                False                 False                False   
687                False                 False                False   
688                False                 False                False   

    Max_Wind_Speed_MPH Mean_Wind_Speed_MPH Max_Gust_Speed_MPH  \
684              False               False               True   
685              False               False               True   
686              False               False               True   
687              False               False              False   
688              False               False               True   

    Precipitation_In Events Mean_Temperature_C  
684            False   True              False  
685            False   True              False  
686            False   True              False  
687            False   True              False  
688            False  False              False  

[5 rows x 22 columns]

In [10]:
# count the number of missing values in each Series
df.isnull().sum()

Date                            0
Max_Temperature_F               0
Mean_Temperature_F              1
Min_TemperatureF                0
Max_Dew_Point_F                 0
MeanDew_Point_F                 0
Min_Dewpoint_F                  0
Max_Humidity                    0
Mean_Humidity                   0
Min_Humidity                    0
Max_Sea_Level_Pressure_In       0
Mean_Sea_Level_Pressure_In      0
Min_Sea_Level_Pressure_In       0
Max_Visibility_Miles            0
Mean_Visibility_Miles           0
Min_Visibility_Miles            0
Max_Wind_Speed_MPH              0
Mean_Wind_Speed_MPH             0
Max_Gust_Speed_MPH            185
Precipitation_In                0
Events                        361
Mean_Temperature_C              1
dtype: int64

# Filling Max_Gust_Speed_MPH  with Some Values

In [11]:
#Checking column for exsisting values
df['Max_Gust_Speed_MPH']

0       21
1       17
2       25
3        -
4        -
5        -
6       18
7        -
8       21
9       22
10      22
11       -
12      41
13      30
14       -
15       -
16      24
17       -
18       -
19       -
20       -
21      20
22       -
23       -
24      37
25       -
26       -
27      16
28       -
29      22
      ... 
659     17
660    NaN
661     21
662    NaN
663    NaN
664    NaN
665    NaN
666    NaN
667     17
668     20
669    NaN
670    NaN
671    NaN
672    NaN
673    NaN
674    NaN
675    NaN
676    NaN
677    NaN
678    NaN
679    NaN
680    NaN
681    NaN
682    NaN
683    NaN
684    NaN
685    NaN
686    NaN
687     18
688    NaN
Name: Max_Gust_Speed_MPH, dtype: object

In [12]:
# fill in missing values with a specified value
df['Max_Gust_Speed_MPH'].fillna(value='0', inplace=True)
df.Max_Gust_Speed_MPH.replace('-','0',inplace=True)
df['Max_Gust_Speed_MPH']

0      21
1      17
2      25
3       0
4       0
5       0
6      18
7       0
8      21
9      22
10     22
11      0
12     41
13     30
14      0
15      0
16     24
17      0
18      0
19      0
20      0
21     20
22      0
23      0
24     37
25      0
26      0
27     16
28      0
29     22
       ..
659    17
660     0
661    21
662     0
663     0
664     0
665     0
666     0
667    17
668    20
669     0
670     0
671     0
672     0
673     0
674     0
675     0
676     0
677     0
678     0
679     0
680     0
681     0
682     0
683     0
684     0
685     0
686     0
687    18
688     0
Name: Max_Gust_Speed_MPH, dtype: object

In [13]:
##Filling Null values with the mean of Max_Gust_Speed_MPH' values
df['Max_Gust_Speed_MPH'] = df['Max_Gust_Speed_MPH'].astype(int)
df['Max_Gust_Speed_MPH'].mean()

9.252539912917271

In [14]:
df.Max_Gust_Speed_MPH.replace('0','9.26',inplace=True)

# Filling Events with Some Values

In [15]:
# count the number of missing values in each Series
df.isnull().sum()

Date                            0
Max_Temperature_F               0
Mean_Temperature_F              1
Min_TemperatureF                0
Max_Dew_Point_F                 0
MeanDew_Point_F                 0
Min_Dewpoint_F                  0
Max_Humidity                    0
Mean_Humidity                   0
Min_Humidity                    0
Max_Sea_Level_Pressure_In       0
Mean_Sea_Level_Pressure_In      0
Min_Sea_Level_Pressure_In       0
Max_Visibility_Miles            0
Mean_Visibility_Miles           0
Min_Visibility_Miles            0
Max_Wind_Speed_MPH              0
Mean_Wind_Speed_MPH             0
Max_Gust_Speed_MPH              0
Precipitation_In                0
Events                        361
Mean_Temperature_C              1
dtype: int64

In [16]:
import pandas as pd
Inplace = True
for index, row in df.iterrows():
    a = row.Precipitation_In
    if a == 0:
        df.at[index, 'Events'] = 'Rain'

In [17]:
# count the number of missing values in each Series
df.isnull().sum()

Date                           0
Max_Temperature_F              0
Mean_Temperature_F             1
Min_TemperatureF               0
Max_Dew_Point_F                0
MeanDew_Point_F                0
Min_Dewpoint_F                 0
Max_Humidity                   0
Mean_Humidity                  0
Min_Humidity                   0
Max_Sea_Level_Pressure_In      0
Mean_Sea_Level_Pressure_In     0
Min_Sea_Level_Pressure_In      0
Max_Visibility_Miles           0
Mean_Visibility_Miles          0
Min_Visibility_Miles           0
Max_Wind_Speed_MPH             0
Mean_Wind_Speed_MPH            0
Max_Gust_Speed_MPH             0
Precipitation_In               0
Events                        14
Mean_Temperature_C             1
dtype: int64

In [18]:
import pandas as pd
Inplace = True
#df['Mean_Temperature_C'] = df['Mean_Temperature_C'].astype(int)
for index, row in df.iterrows():
    a = row.Mean_Temperature_C
    if a >= 14:
        df.at[index, 'Events'] = 'Sunny'
    print (a)

16.67
15.0
14.44
16.11
15.56
17.78
17.78
15.56
14.44
14.44
12.78
13.33
14.44
11.11
12.22
13.33
15.0
14.44
12.22
11.11
10.56
14.44
14.44
13.89
15.56
11.11
8.89
11.11
8.89
5.56
3.89
4.44
3.33
3.33
3.89
5.0
3.89
6.67
9.44
11.11
8.89
10.0
8.33
13.33
15.0
14.44
8.89
1.67
0.56
1.11
1.67
4.44
6.67
10.0
11.11
9.44
12.78
13.89
16.11
12.78
11.11
7.78
8.33
8.89
8.89
8.89
8.89
10.0
11.11
12.78
8.89
8.89
6.67
5.56
5.56
8.33
6.67
5.56
1.11
1.11
1.67
3.33
3.33
8.89
13.33
10.0
8.33
6.67
6.67
7.78
8.89
9.44
6.67
5.56
5.0
10.0
7.78
12.22
9.44
7.78
4.44
8.89
12.22
14.44
13.89
12.78
11.11
10.0
8.89
6.67
5.56
8.89
10.0
8.89
8.33
12.22
13.33
12.22
12.22
11.11
11.11
10.56
16.67
12.22
12.22
8.89
11.67
11.11
8.89
10.56
11.11
10.0
7.22
6.67
7.22
9.44
10.56
10.56
8.33
6.67
9.44
5.56
6.67
8.89
9.44
10.0
10.0
9.44
10.0
13.33
15.56
15.0
14.44
10.0
11.11
10.56
12.22
13.33
13.33
12.78
9.44
8.89
11.11
12.22
16.67
15.0
14.44
13.89
15.56
10.0
10.0
10.0
9.44
8.89
10.0
11.11
12.22
12.22
12.22
12.22
10.0
11.11
8.89
8.33
9.

In [1]:
#Search google for Sunny condition in Seattle
from IPython.display import HTML
HTML('<iframe src=https://www.accuweather.com/en/us/seattle-wa/98104/current-weather/351409 width=1000 height=350></iframe>')

In [20]:
# count the number of missing values in each Series
df.isnull().sum()

Date                          0
Max_Temperature_F             0
Mean_Temperature_F            1
Min_TemperatureF              0
Max_Dew_Point_F               0
MeanDew_Point_F               0
Min_Dewpoint_F                0
Max_Humidity                  0
Mean_Humidity                 0
Min_Humidity                  0
Max_Sea_Level_Pressure_In     0
Mean_Sea_Level_Pressure_In    0
Min_Sea_Level_Pressure_In     0
Max_Visibility_Miles          0
Mean_Visibility_Miles         0
Min_Visibility_Miles          0
Max_Wind_Speed_MPH            0
Mean_Wind_Speed_MPH           0
Max_Gust_Speed_MPH            0
Precipitation_In              0
Events                        7
Mean_Temperature_C            1
dtype: int64

In [21]:
# confirm that the missing values were filled in
df['Events'].value_counts().head()

Rain                 339
Sunny                325
Fog , Rain             5
Fog-Rain               5
Rain-Thunderstorm      3
Name: Events, dtype: int64

# Encoding Columns values

In [22]:
df.Events.replace('Fog-Rain', 'Fog, Rain',inplace=True)
df.Events.replace('Fog , Rain', 'Fog, Rain',inplace=True)

In [23]:
df['Events'].value_counts().head()

Rain                 339
Sunny                325
Fog, Rain             10
Rain-Thunderstorm      3
Fog                    2
Name: Events, dtype: int64

# Converting Categorical variable to nominal 

In [24]:
df["Events_cat"] = df["Events"].astype('category')
df.dtypes 

Date                            object
Max_Temperature_F                int64
Mean_Temperature_F             float64
Min_TemperatureF                 int64
Max_Dew_Point_F                  int64
MeanDew_Point_F                  int64
Min_Dewpoint_F                   int64
Max_Humidity                     int64
Mean_Humidity                    int64
Min_Humidity                     int64
Max_Sea_Level_Pressure_In      float64
Mean_Sea_Level_Pressure_In     float64
Min_Sea_Level_Pressure_In      float64
Max_Visibility_Miles             int64
Mean_Visibility_Miles            int64
Min_Visibility_Miles             int64
Max_Wind_Speed_MPH               int64
Mean_Wind_Speed_MPH              int64
Max_Gust_Speed_MPH              object
Precipitation_In               float64
Events                          object
Mean_Temperature_C             float64
Events_cat                    category
dtype: object

In [25]:
df["Events_num"] = df["Events_cat"].cat.codes

In [29]:
df["Events_cat"] = df["Events"].astype('category')
df["Events_num"] = df["Events_cat"].cat.codes

In [36]:
df['Events_num'].mode()

0    2
dtype: int8

In [37]:
df[(df.Events_num == 2)]

Date  Max_Temperature_F  Mean_Temperature_F  Min_TemperatureF  \
10   10/23/2014                 62                55.0                50   
11   10/24/2014                 60                56.0                51   
13   10/26/2014                 59                52.0                48   
14   10/27/2014                 62                54.0                45   
15   10/28/2014                 62                56.0                51   
18   10/31/2014                 59                54.0                52   
19   11/01/2014                 55                52.0                48   
20   11/02/2014                 57                51.0                45   
23   11/05/2014                 61                57.0                53   
25   11/07/2014                 59                52.0                45   
26   11/08/2014                 55                48.0                42   
27   11/09/2014                 57                52.0                48   
28   11/10/2014                 54                48.0                43   
29   11/11/2014                 48                42.0                36   
30   11/12/2014                 46                39.0                32   
31   11/13/2014                 48                40.0                30   
32   11/14/2014                 48                38.0                28   
33   11/15/2014                 48                38.0                27   
34   11/16/2014                 50                39.0                28   
35   11/17/2014                 52                41.0                30   
36   11/18/2014                 48                39.0                30   
37   11/19/2014                 52                44.0                36   
38   11/20/2014                 54                49.0                44   
39   11/21/2014                 55                52.0                50   
40   11/22/2014                 51                48.0                45   
41   11/23/2014                 57                50.0                44   
42   11/24/2014                 55                47.0                39   
43   11/25/2014                 60                56.0                51   
46   11/28/2014                 57                48.0                39   
48   11/30/2014                 39                33.0                27   
..          ...                ...                 ...               ...   
533   3/29/2016                 64                50.0                37   
534   3/30/2016                 68                54.0                43   
537  04/02/2016                 63                54.0                48   
539  04/04/2016                 60                54.0                46   
540  04/05/2016                 57                52.0                46   
541  04/06/2016                 68                56.0                45   
544  04/09/2016                 66                55.0                46   
545  04/10/2016                 59                52.0                48   
546  04/11/2016                 57                54.0                50   
547  04/12/2016                 60                52.0                46   
548   4/13/2016                 59                52.0                46   
549   4/14/2016                 59                52.0                45   
550   4/15/2016                 60                54.0                48   
551   4/16/2016                 68                56.0                43   
559   4/24/2016                 55                51.0                48   
561   4/26/2016                 60                52.0                43   
562   4/27/2016                 61                56.0                51   
564   4/29/2016                 61                55.0                50   
565   4/30/2016                 68                56.0                45   
569  05/04/2016                 60                56.0                54   
573  05/08/2016                 63                56.0     

As we can see that rain is our mean Event We will fill all empty event with 6

In [38]:
# fill in missing values with a specified value
df['Events'].fillna(value='Rain', inplace=True)

The Events_num is appera as -1 which is not right, to fix this problem we repeate above step

In [39]:
df["Events_cat"] = df["Events"].astype('category')
df["Events_num"] = df["Events_cat"].cat.codes

In [40]:
# count the number of missing values in each Series
df.isnull().sum()

Date                          0
Max_Temperature_F             0
Mean_Temperature_F            1
Min_TemperatureF              0
Max_Dew_Point_F               0
MeanDew_Point_F               0
Min_Dewpoint_F                0
Max_Humidity                  0
Mean_Humidity                 0
Min_Humidity                  0
Max_Sea_Level_Pressure_In     0
Mean_Sea_Level_Pressure_In    0
Min_Sea_Level_Pressure_In     0
Max_Visibility_Miles          0
Mean_Visibility_Miles         0
Min_Visibility_Miles          0
Max_Wind_Speed_MPH            0
Mean_Wind_Speed_MPH           0
Max_Gust_Speed_MPH            0
Precipitation_In              0
Events                        0
Mean_Temperature_C            1
Events_cat                    0
Events_num                    0
dtype: int64

In [41]:
# confirm that the missing values were filled in
df['Events'].value_counts().head()

Rain                 346
Sunny                325
Fog, Rain             10
Rain-Thunderstorm      3
Fog                    2
Name: Events, dtype: int64

In [42]:
# if 'any' values are missing in a row, then drop that row
df.dropna(how='any').shape

(688, 24)

In [43]:
# count the number of missing values in each Series
df.isnull().sum()

Date                          0
Max_Temperature_F             0
Mean_Temperature_F            1
Min_TemperatureF              0
Max_Dew_Point_F               0
MeanDew_Point_F               0
Min_Dewpoint_F                0
Max_Humidity                  0
Mean_Humidity                 0
Min_Humidity                  0
Max_Sea_Level_Pressure_In     0
Mean_Sea_Level_Pressure_In    0
Min_Sea_Level_Pressure_In     0
Max_Visibility_Miles          0
Mean_Visibility_Miles         0
Min_Visibility_Miles          0
Max_Wind_Speed_MPH            0
Mean_Wind_Speed_MPH           0
Max_Gust_Speed_MPH            0
Precipitation_In              0
Events                        0
Mean_Temperature_C            1
Events_cat                    0
Events_num                    0
dtype: int64

In [44]:
# fill in missing values with a specified value
df['Mean_Temperature_F'].fillna(value='57', inplace=True)

In [45]:
# count the number of missing values in each Series
df.isnull().sum()

Date                          0
Max_Temperature_F             0
Mean_Temperature_F            0
Min_TemperatureF              0
Max_Dew_Point_F               0
MeanDew_Point_F               0
Min_Dewpoint_F                0
Max_Humidity                  0
Mean_Humidity                 0
Min_Humidity                  0
Max_Sea_Level_Pressure_In     0
Mean_Sea_Level_Pressure_In    0
Min_Sea_Level_Pressure_In     0
Max_Visibility_Miles          0
Mean_Visibility_Miles         0
Min_Visibility_Miles          0
Max_Wind_Speed_MPH            0
Mean_Wind_Speed_MPH           0
Max_Gust_Speed_MPH            0
Precipitation_In              0
Events                        0
Mean_Temperature_C            1
Events_cat                    0
Events_num                    0
dtype: int64

In [46]:
df['Mean_Temperature_C'].mean()

13.657965116279073

In [47]:
df['Mean_Temperature_C'].fillna(value='14', inplace=True)

In [48]:
# count the number of missing values in each Series
df.isnull().sum()

Date                          0
Max_Temperature_F             0
Mean_Temperature_F            0
Min_TemperatureF              0
Max_Dew_Point_F               0
MeanDew_Point_F               0
Min_Dewpoint_F                0
Max_Humidity                  0
Mean_Humidity                 0
Min_Humidity                  0
Max_Sea_Level_Pressure_In     0
Mean_Sea_Level_Pressure_In    0
Min_Sea_Level_Pressure_In     0
Max_Visibility_Miles          0
Mean_Visibility_Miles         0
Min_Visibility_Miles          0
Max_Wind_Speed_MPH            0
Mean_Wind_Speed_MPH           0
Max_Gust_Speed_MPH            0
Precipitation_In              0
Events                        0
Mean_Temperature_C            0
Events_cat                    0
Events_num                    0
dtype: int64

In [49]:
df['Mean_Temperature_F'] = df['Mean_Temperature_F'].astype(int)
df['Mean_Temperature_F'].mean()

56.58490566037736

In [50]:
# 'isnull' returns a DataFrame of booleans (True if missing, False if not missing)
df.isnull().tail()

Date Max_Temperature_F Mean_Temperature_F Min_TemperatureF  \
684  False             False              False            False   
685  False             False              False            False   
686  False             False              False            False   
687  False             False              False            False   
688  False             False              False            False   

    Max_Dew_Point_F MeanDew_Point_F Min_Dewpoint_F Max_Humidity Mean_Humidity  \
684           False           False          False        False         False   
685           False           False          False        False         False   
686           False           False          False        False         False   
687           False           False          False        False         False   
688           False           False          False        False         False   

    Min_Humidity    ...     Mean_Visibility_Miles Min_Visibility_Miles  \
684        False    ...                     False                False   
685        False    ...                     False                False   
686        False    ...                     False                False   
687        False    ...                     False                False   
688        False    ...                     False                False   

    Max_Wind_Speed_MPH Mean_Wind_Speed_MPH Max_Gust_Speed_MPH  \
684              False               False              False   
685              False               False              False   
686              False               False              False   
687              False               False              False   
688              False               False              False   

    Precipitation_In Events Mean_Temperature_C Events_cat Events_num  
684            False  False              False      False      False  
685            False  False              False      False      False  
686            False  False              False      False      False  
687            False  False              False      False      False  
688            False  False              False      False      False  

[5 rows x 24 columns]

# Encoding Data

In [51]:
# convert 'Time' to datetime format
df['Date'] = pd.to_datetime(df.Date)

In [52]:
df['month'] = df['Date'].dt.month

In [53]:
# remove a single column (axis=1 refers to columns)
df.drop('Events_cat', axis=1, inplace=True)

In [54]:
df['year']=df['Date'].dt.year

# Removing Duplicated Tuples

In [55]:
# detect duplicate Date codes: True if an item is identical to a previous item
df.Date.duplicated().tail()

684    False
685    False
686    False
687    False
688    False
Name: Date, dtype: bool

In [56]:
# count the duplicate items (True becomes 1, False becomes 0)
df.Date.duplicated().sum()

0

The above codes shows that we do not have any duplicated rows in our dataset

In [57]:
df.head(100)

Date  Max_Temperature_F  Mean_Temperature_F  Min_TemperatureF  \
0  2014-10-13                 71                  62                54   
1  2014-10-14                 63                  59                55   
2  2014-10-15                 62                  58                54   
3  2014-10-16                 71                  61                52   
4  2014-10-17                 64                  60                57   
5  2014-10-18                 68                  64                59   
6  2014-10-19                 73                  64                55   
7  2014-10-20                 66                  60                55   
8  2014-10-21                 64                  58                55   
9  2014-10-22                 60                  58                57   
10 2014-10-23                 62                  55                50   
11 2014-10-24                 60                  56                51   
12 2014-10-25                 64                  58                52   
13 2014-10-26                 59                  52                48   
14 2014-10-27                 62                  54                45   
15 2014-10-28                 62                  56                51   
16 2014-10-29                 64                  59                54   
17 2014-10-30                 62                  58                55   
18 2014-10-31                 59                  54                52   
19 2014-11-01                 55                  52                48   
20 2014-11-02                 57                  51                45   
21 2014-11-03                 60                  58                55   
22 2014-11-04                 61                  58                55   
23 2014-11-05                 61                  57                53   
24 2014-11-06                 64                  60                54   
25 2014-11-07                 59                  52                45   
26 2014-11-08                 55                  48                42   
27 2014-11-09                 57                  52                48   
28 2014-11-10                 54                  48                43   
29 2014-11-11                 48                  42                36   
..        ...                ...                 ...               ...   
70 2014-12-22                 54                  48                42   
71 2014-12-23                 53                  48                43   
72 2014-12-24                 46                  44                41   
73 2014-12-25                 48                  42                37   
74 2014-12-26                 44                  42                39   
75 2014-12-27                 52                  47                42   
76 2014-12-28                 46                  44                41   
77 2014-12-29                 46                  42                35   
78 2014-12-30                 39                  34                30   
79 2014-12-31                 41                  34                26   
80 2015-01-01                 43                  35                27   
81 2015-01-02                 44                  38                32   
82 2015-01-03                 43                  38                33   
83 2015-01-04                 54                  48                41   
84 2015-01-05                 57                  56                54   
85 2015-01-06                 55                  50                46   
86 2015-01-07                 50                  47                44   
87 2015-01-08                 48                  44                39   
88 2015-01-09                 50                  44                39   
89 2015-01-10                 48                  46                44   
90 2015-01-11                 51                  48                46   
91 2015-01-12                 53                  49                45   
92 2015-01-13

In [58]:
df['year'] = df['year'].astype(int)

In [59]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 689 entries, 0 to 688
Data columns (total 25 columns):
Date                          689 non-null datetime64[ns]
Max_Temperature_F             689 non-null int64
Mean_Temperature_F            689 non-null int32
Min_TemperatureF              689 non-null int64
Max_Dew_Point_F               689 non-null int64
MeanDew_Point_F               689 non-null int64
Min_Dewpoint_F                689 non-null int64
Max_Humidity                  689 non-null int64
Mean_Humidity                 689 non-null int64
Min_Humidity                  689 non-null int64
Max_Sea_Level_Pressure_In     689 non-null float64
Mean_Sea_Level_Pressure_In    689 non-null float64
Min_Sea_Level_Pressure_In     689 non-null float64
Max_Visibility_Miles          689 non-null int64
Mean_Visibility_Miles         689 non-null int64
Min_Visibility_Miles          689 non-null int64
Max_Wind_Speed_MPH            689 non-null int64
Mean_Wind_Speed_MPH           689 non-null int6

In [60]:
df.Date.tail()

684   2016-08-27
685   2016-08-28
686   2016-08-29
687   2016-08-30
688   2016-08-31
Name: Date, dtype: datetime64[ns]

In [61]:
# df1 belong to weather dataset
df1 = df[(df.year != 2014)]
df1.tail()

Date  Max_Temperature_F  Mean_Temperature_F  Min_TemperatureF  \
684 2016-08-27                 72                  66                61   
685 2016-08-28                 75                  68                59   
686 2016-08-29                 81                  68                55   
687 2016-08-30                 70                  64                57   
688 2016-08-31                 71                  65                59   

     Max_Dew_Point_F  MeanDew_Point_F  Min_Dewpoint_F  Max_Humidity  \
684               57               54              50            81   
685               54               52              50            80   
686               55               53              50            89   
687               55               53              52            83   
688               61               56              52            90   

     Mean_Humidity  Min_Humidity  ...   Min_Visibility_Miles  \
684             65            46  ...                     10   
685             65            44  ...                     10   
686             65            39  ...                      6   
687             69            53  ...                     10   
688             77            63  ...                      8   

     Max_Wind_Speed_MPH  Mean_Wind_Speed_MPH  Max_Gust_Speed_MPH  \
684                  16                    9                9.26   
685                  12                    9                9.26   
686                   9                    4                9.26   
687                  14                    9                  18   
688                  14                    8                9.26   

     Precipitation_In  Events  Mean_Temperature_C  Events_num month  year  
684               0.0   Sunny               18.89           6     8  2016  
685               0.0   Sunny                  20           6     8  2016  
686               0.0   Sunny                  20           6     8  2016  
687               0.0   Sunny               17.78           6     8  2016  
688               0.0   Sunny               18.33           6     8  2016  

[5 rows x 25 columns]

# Saving Weather Data

Weather data consist on 600 observations where bikes data is consisting over a hundred thousand of tuples which makles it hard to combine both data set for further weather analysis.  To achieve this we will, break both weather and bikes data Date columns and applying programming technique to achieve weather value for each bicycle trip.

In [62]:
df1.isnull().sum()

Date                          0
Max_Temperature_F             0
Mean_Temperature_F            0
Min_TemperatureF              0
Max_Dew_Point_F               0
MeanDew_Point_F               0
Min_Dewpoint_F                0
Max_Humidity                  0
Mean_Humidity                 0
Min_Humidity                  0
Max_Sea_Level_Pressure_In     0
Mean_Sea_Level_Pressure_In    0
Min_Sea_Level_Pressure_In     0
Max_Visibility_Miles          0
Mean_Visibility_Miles         0
Min_Visibility_Miles          0
Max_Wind_Speed_MPH            0
Mean_Wind_Speed_MPH           0
Max_Gust_Speed_MPH            0
Precipitation_In              0
Events                        0
Mean_Temperature_C            0
Events_num                    0
month                         0
year                          0
dtype: int64

# Data type Conversion

In [63]:
df1['Mean_Temperature_C'] = df1['Mean_Temperature_C'].astype(int)
df1.dtypes

Date                          datetime64[ns]
Max_Temperature_F                      int64
Mean_Temperature_F                     int32
Min_TemperatureF                       int64
Max_Dew_Point_F                        int64
MeanDew_Point_F                        int64
Min_Dewpoint_F                         int64
Max_Humidity                           int64
Mean_Humidity                          int64
Min_Humidity                           int64
Max_Sea_Level_Pressure_In            float64
Mean_Sea_Level_Pressure_In           float64
Min_Sea_Level_Pressure_In            float64
Max_Visibility_Miles                   int64
Mean_Visibility_Miles                  int64
Min_Visibility_Miles                   int64
Max_Wind_Speed_MPH                     int64
Mean_Wind_Speed_MPH                    int64
Max_Gust_Speed_MPH                    object
Precipitation_In                     float64
Events                                object
Mean_Temperature_C                     int32
Events_num

In [64]:
df1

Date  Max_Temperature_F  Mean_Temperature_F  Min_TemperatureF  \
80  2015-01-01                 43                  35                27   
81  2015-01-02                 44                  38                32   
82  2015-01-03                 43                  38                33   
83  2015-01-04                 54                  48                41   
84  2015-01-05                 57                  56                54   
85  2015-01-06                 55                  50                46   
86  2015-01-07                 50                  47                44   
87  2015-01-08                 48                  44                39   
88  2015-01-09                 50                  44                39   
89  2015-01-10                 48                  46                44   
90  2015-01-11                 51                  48                46   
91  2015-01-12                 53                  49                45   
92  2015-01-13                 48                  44                41   
93  2015-01-14                 46                  42                37   
94  2015-01-15                 46                  41                36   
95  2015-01-16                 55                  50                44   
96  2015-01-17                 55                  46                39   
97  2015-01-18                 60                  54                48   
98  2015-01-19                 52                  49                46   
99  2015-01-20                 51                  46                39   
100 2015-01-21                 48                  40                33   
101 2015-01-22                 52                  48                45   
102 2015-01-23                 57                  54                50   
103 2015-01-24                 62                  58                55   
104 2015-01-25                 64                  57                50   
105 2015-01-26                 64                  55                46   
106 2015-01-27                 55                  52                50   
107 2015-01-28                 55                  50                46   
108 2015-01-29                 55                  48                42   
109 2015-01-30                 48                  44                39   
..         ...                ...                 ...               ...   
659 2016-08-02                 73                  64                57   
660 2016-08-03                 75                  66                60   
661 2016-08-04                 80                  69                59   
662 2016-08-05                 79                  70                61   
663 2016-08-06                 72                  63                55   
664 2016-08-07                 72                  66                60   
665 2016-08-08                 73                  64                57   
666 2016-08-09                 72                  66                60   
667 2016-08-10                 75                  68                60   
668 2016-08-11                 80                  70                61   
669 2016-08-12                 87                  75                63   
670 2016-08-13                 90                  78                66   
671 2016-08-14                 82                  71                60   
672 2016-08-15                 84                  72                61   
673 2016-08-16                 81                  69                57   
674 2016-08-17                 79                  70                60   
675 2016-08-18                 86                  73                60   
676 2016-08-19                 95                  60                26   
677 2016-08-20                 91                  78                64   
678 2016-08-21                 73                  66                60   
679 2016-08-22                 72                  64                57   
680 2016-08-23                 79  

In [65]:
df1.to_csv('C:\\Users\\mrferozi\\Desktop\\GitHub\\Bike\\dataset\\cycle\\weather_clean.csv')

## Resources
References:*From the video series: [Introduction to machine learning with scikit-learn](https://github.com/justmarkham/scikit-learn-videos)*
- scikit-learn documentation: [Cross-validation](http://scikit-learn.org/stable/modules/cross_validation.html), [Model evaluation](http://scikit-learn.org/stable/modules/model_evaluation.html)
- scikit-learn issue on GitHub: [MSE is negative when returned by cross_val_score](https://github.com/scikit-learn/scikit-learn/issues/2439)
- Section 5.1 of [An Introduction to Statistical Learning](http://www-bcf.usc.edu/~gareth/ISL/) (11 pages) and related videos: [K-fold and leave-one-out cross-validation](https://www.youtube.com/watch?v=nZAM5OXrktY) (14 minutes), [Cross-validation the right and wrong ways](https://www.youtube.com/watch?v=S06JpVoNaA0) (10 minutes)
- Scott Fortmann-Roe: [Accurately Measuring Model Prediction Error](http://scott.fortmann-roe.com/docs/MeasuringError.html)
- Machine Learning Mastery: [An Introduction to Feature Selection](http://machinelearningmastery.com/an-introduction-to-feature-selection/)
- Harvard CS109: [Cross-Validation: The Right and Wrong Way](https://github.com/cs109/content/blob/master/lec_10_cross_val.ipynb)
- Journal of Cheminformatics: [Cross-validation pitfalls when selecting and assessing regression and classification models](http://www.jcheminf.com/content/pdf/1758-2946-6-10.pdf)